д/з 3 — Жевак Антон

mayton opened the pull request:
Вроде всё работает :) Из особенностей:
- возможность выбирать количество слов в префиксе
- автоматический поиск и вставка КДПВ
- (возможная) вставка в статью программного кода на языке, который в ней упоминается 
- дополнительная обработка входного текста с помощью регулярных выражений

https://rawgithub.com/mayton/3-oop-js-dom/master/index.html

mayton committed:
вполне себе рабочий вариант

mayton committed:
раскомментировал вставку картинки

mayton committed:
попытка исправить вставку картинки

mayton committed:
удалено лишнее, улучшен эстетический вид

mayton committed:
исправлена опечатка в имени массива

mayton committed:
ни на что не влияющие улучшения

VladimirShestakov commented:
Такого бреда я ещё не читал =)) Комменты иерархией ещё сделай

NiceJuli commented:
Бредовый бред :D

mayton commented:
Теперь я смогу писать крутые любовные письма!

mayton committed:
сжал хабра-стили в одну css

mayton committed:
удалены неиспользуемые стили для модального окна

ultd39 commented:
Первая мысль — «Зачем меня средиректило на хабр?». А спустя секунду все встало на свои места. :)

Классная работа :)

ivanzolotov commented:
Прекрасная работа. Предела совершенству, естественно, нет, но и сейчас это вполне достойный генератор первоклассного бреда.

azproduction commented:
Индексацию текста можно вынести в WebWorker. Сейчас анализ текста занимает много времени.

mayton committed:
обновлён habr.html

теперь при валидации не возникают unclosed element и stray end tag

maxvipon commented:
1. Хорошо, что сжал стили, но по-хорошему сперва стоило выпилить неиспользуемые
2. Обязательно нужно сообщать пользователю, что процесс идет, что все работает и станица не повисла. Сейчас после нажатия кнопки в модальном окне непонятно, работает ли оно.
3. Зачем вообще нужно модальное окно? Можно ли его как-то вызвать уже на сгенерированной страничке?

azproduction commented:
> Зачем вообще нужно модальное окно? Можно ли его как-то вызвать уже на сгенерированной страничке?

Помоему тут все логично.

mayton commented:
@azproduction, а что под индексацией понимается? В Web Workers нет window и document, значит jQuery там не работает, так что гулять по DOM-элементам, выдирать оттуда текст и фильтровать его не получится. По крайней мере у меня не получилось :-) Я попытался вынести в worker создание словаря, но по-моему стало ещё медленнее :-) так что коммитить не стал.

Поэтому пока особого смысла сообщать пользователю о идущем процессе, наверное, нет. Ведь в это время даже gif-индикатор крутиться не будет.

А в модальном окне задаются настройки генерируемого текста, так что оно как бы да, показывается перед его генерацией.

maxvipon commented:
> А в модальном окне задаются настройки генерируемого текста, так что оно как бы да, показывается перед его генерацией.

Спасибо, Кэп!

Я убираю через DevConsole модальное окно и оверлай и вижу – тадам! – пустую белую страницу. Отсюда вопрос: Почему форма с настройками вынесена в псевдомодальное окошко, а не сделана просто на станице?

> Поэтому пока особого смысла сообщать пользователю о идущем процессе, наверное, нет.

Вот если бы разработчики интерфейсов, например, банкоматов думали также, было бы весело.

В общем, после нажатия кнопки «Поехали», нужно как минимум показать текст «Идет генерация…»

azproduction commented:
> что под индексацией понимается?

Построение структуры под цепь Маркова (создание словаря).

Ты знаешь что у тебя тормозит? Сбор DOM-нод или создание словаря или что-то еще? Давай найдем узкое место и будем оптимизировать.

PS Заставить индикатор крутиться в однопоточной обработке можно. Достаточно делать таймаут раз в 50-100 мс, чтобы фрейм, который перерисовывает интерфейс отработал.

mayton committed:
полностью переработан алгоритм, много всего

полностью переработан алгоритм разбора
теперь используется web workers
добавлен индикатор текущего состояния
добавлен показ времени генерации и т.д.

mayton committed:
вставку картинки забыл

mayton commented:
@maxvipon, добавил индикатор текущего состояния. Но за что ты так возненавидел модальное окно, не понимаю. Оно у меня готовое было, вот и приделал его, чтоб стили не писать.

@azproduction, тормозило объединение массивов. Сейчас я алгоритм переписал: теперь все комментарии сначала объединяются в один текст, а только потом разбираются на слова. Статьи аналогично. Работнику доверил подготовку текста и составление словарей.

maxvipon commented:
> Но за что ты так возненавидел модальное окно, не понимаю.

Не плоди сущности без необходимости ©

Модальные окна вообще в интерфейсе нужно применять с большой осторожностью.
Это давняя и больная тема для UX.

mayton committed:
древовидные комментарии, убрал модальное окно

mayton committed:
отключаем кнопку при нажатии

mayton committed:
теперь достаточно 1 раз сгенерировать словарь

azproduction commented:
:white_check_mark: Все хорошо. Есть мелочи по кодстайлу, но они минорные.
Ну вот с воркерами даже не лагает! Хорошо, что менюха с настройками генератора ушла в родной интерфейс хабра.
